src/libponyc/pass/docgen.c

```diff
@@ -548,6 +548,89 @@ static void doc_type_params(docgen_t* docgen, ast_t* t_params,
     fprintf(docgen->type_file, "]");
 }
 
+// Write the expression to the current type file.
+void doc_expression(docgen_t* docgen, ast_t* ast)
+{
+  FILE* fp = docgen->type_file;
+  ast_t* child = ast_child(ast);
+
+  switch(ast_id(ast))
+  {
+    case TK_CALL:
+      while(child != NULL)
+      {
+        doc_expression(docgen, child);
+        child = ast_sibling(child);
+      }
+      fprintf(fp, "()");
+      break;
+
+    case TK_SEQ:
+      while(child != NULL)
+      {
+        doc_expression(docgen, child);
+        child = ast_sibling(child);
+        if (child != NULL) {
+          fprintf(fp, "; ");
+        }
+      }
+      break;
+
+    case TK_REFERENCE:
+      doc_expression(docgen, child);
+      break;
+
+    case TK_TYPE:
+    case TK_INTERFACE:
+    case TK_TRAIT:
+    case TK_PRIMITIVE:
+    case TK_STRUCT:
+    case TK_CLASS:
+    case TK_ACTOR:
+      doc_type(docgen, ast, false);
+      break;
+
+    case TK_DOT:
```

Comment:
This could be combined with TK_SEQ and TK_CALL and others that are the same no? There's a slight variance in character that would have to be handled but there's a ton of duplication and it would be easy to introduce bugs in this.